### DISCUSSION TITLE
Improving x86 emulation

### DISCUSSION THREAD

**Discussion by nehochupechatat** [danoon2/Boxedwine#70] [Feature Requests]
It seems like the x86 emulator is very simple, so it runs at a level of a 90s Pentium 1.

My suggestion is to either use FEX (a very optimized Linux-target x86 HLE with both 32 and 64bit support), or Box86/64 for ARM targets, and removing CPU emulation from the Windows build so it runs kinda like a container.

**Comment by danoon2** [comment]
Thanks for the FEX mention.  I looked into that.  They use LLVM to handle the ASM translations.  

I also notice they will thunk Vulkan.  I'm currently try to get that to work in Boxedwine so I will definitely take a look to see how FEX handles that.

As for performance, what platform are your referring to for a 90s Pentium 1 performance?  If you look at the front page, https://github.com/danoon2/Boxedwine and scroll down, you will see I put up some performance numbers.  The new M1 Arm chip is about 20x faster than a Pentium 90 on the MDK performance test.  And that is about 2x faster than what QEMU can do.  I'm not sure how to compare it FEX unless someone got FEX to run Wine.

As for the Windows build, the x64 build uses full native translation of x86 to x64.  It should run at about 25% of the native CPU.

One of the challenges for Boxedwine is that I choose to emulate processes inside a single process.  So each process need to emulate the MMU.  FEX doesn't have to do that which simplifies things a lot.

**Comment by nehochupechatat** [comment]
Yes, Wine has been shown to work on FEX. It can actually run DOOM (2016) at 30fps on a SD888! Pretty amazing stuff. You could check their YouTube channel.

As for the Windows port, I suggest decoupling it from the emulation code entirely since it's all x86 and just using a simple container.

**Comment by nehochupechatat** [comment]
https://youtu.be/aU9cWW6rpx4

Here is a demo of FEX + Wine on a SD888 chip.

**Comment by danoon2** [comment]
The x64 Windows build of Boxedwine is already pretty fast, faster than any of the games I'm currently working on, which are games before the year 2000 or so.  But one possibility for improved performance if I need it would be to support native chip virtualization of the MMU.  The 16-bit and 32-bit code both get converted to native x64 so it should work.  This would require me to write a windows driver or maybe use Intel HAXM

The approach FEX takes wouldn't work for me just because I emulate processes inside a single host process.  So when a wine client (game) talks to the wineserver in another process using unix sockets in Linux, in Boxedwine I just use queues to emulate the pipe/socket and each process is contained in a single host process where I emulated the MMU.  This is why Boxedwine works on things like empscripten/wasm in the browser.

Using LLVM like FEX wouldn't increase the x86/x64 performance, but when converting x86 to ARMv8, I'm sure LLVM does a better job than I do.  In the future I would like to measure the performance and see.  If its a lot faster, I would consider and LLVM backend for the CPU.

**Reply by duckcantcode** [reply]
What about box86?

**Reply by danoon2** [reply]
box86 emulates the cpu like Boxedwine, but for certain dlls, it can forward the calls to a native version.  This will only work on a system that supports 32-bit executables.   For Boxedwine, I was really aiming for run anywhere.

**Reply by duckcantcode** [reply]
We are talking about running box86 within Boxedwine right? Wouldn't the 32-bit executables be run within Boxedwine?

**Reply by danoon2** [reply]
I thought you meant seeing if the approach box86 uses could be applied to Boxedwine. :)

As for running box86 inside Boxedwine, that would be an interesting test of Boxedwine, but I don't think it would achieve any additional benefit.  Whatever bugs Boxedwine has, box86 inside Boxedwine would have the same issues since it runs its system dll's natively just like Wine in Boxedwine does.  But I would expect the performance to be pretty bad.

**Reply by duckcantcode** [reply]
> I thought you meant seeing if the approach box86 uses could be applied to Boxedwine. :)

Ah, I see. Though it would be interesting to directly forward _graphics_ calls from OpenGL and Vulkan, it doesn't make much sense to forward things like SDL2 to have SDL2 games make a window directly on the host system.

**Reply by danoon2** [reply]
I do forward the graphics calls, but my emulation memory space and the memory space of the host process that Boxedwine runs in are separate so I have to marshal the data in and out.  Marshaling a pointer out is usually pretty easy, I can just pass the raw pointer that represents the emulated memory.  But going the other way is really tough.  To make the host/os/opengl pointer visible in emulation takes a lot of work and often times I just copy the data if its appropriate.  Now going from 32-bit emulation to 64-bit host/app is where there can get really interesting because the pointer sizes are different. :)

**Reply by duckcantcode** [reply]
Thank you for giving such a detailed explanation!

What about D3D/DirectX? I see a bunch of D3D related lines in logs, are those calls forwarded too? On Linux hosts that wouldn't work because there's no D3D on Linux (aside from DXVK; would that help us in Boxedwine?)

When setting the renderer, does that change what happens to these forwarded calls?

These topics interest me so I have asked a lot of questions here and in Issues, apologies for asking so many.

**Reply by danoon2** [reply]
Wine will automatically convert DirectX/3D to Opengl and dxvk will do Directx/3D to Vulkan.  So I need to handle forwarding OpenGL from 32-bit Wine/Linux to the host OpenGL(32-bit or 64-bit).  I'm also about half way done with handling the Vulkan forwarding.

The interesting rendering option for Wine is GDI, it will translate DirectDraw (not 3d) to just drawing on a buffer rather than forwarding to OpenGL.  This makes some old DirectDraw games that mix GDI controls (text boxes etc) with DirectDraw work better, like Diablo 1.

**Comment by nehochupechatat** [comment]
What about removing the emulation component entirely for the Windows build? I was thinking of using it as a lightweight Windows container that doesn't take the weight of a Windows installation. I ran Flash MX (the animation software, not the player) on it. Tons of rendering bugs, crashes, and very bad performance even on the 64bit version. This doesn't happen on a regular Linux VM with Wine, so I assume it's the emulation aspect.

**Reply by danoon2** [reply]
I agree with what you are saying, using Boxedwine as a container in Windows is a good idea.  But Boxedwine is probably a long way away from that.  And I doubt it would ever be a perfect container (something more like Sandboxie) since I will always use Wine.  I still have lots of bugs and lots of things to do.  My current priority is working on older games since current Windows versions have trouble or can't run at all the older things, like 16-bit games/installers.

As for Flash MX, since it runs on Wine it should work on Boxedwine at some point.  I haven't looked at it yet.   That might be something interesting to look into though, perhaps later next year.
